File: Python/KeywordRetriever.py
from typing import List, TYPE_CHECKING
from .Hit import Hit
import re

if TYPE_CHECKING:
    from .Chunk import Chunk

def _normalize(text: str) -> str:
    # Türkçe karakterleri ve büyük harfleri normalize et
    text = text.lower().replace('ı', 'i').replace('ş', 's').replace('ğ', 'g').replace('ü', 'u').replace('ö', 'o').replace('ç', 'c')
    return "".join(c for c in text if c.isalnum() or c.isspace())

def retrieve_keywords(terms: List[str], corpus: List['Chunk']) -> List[Hit]:
    hits: List[Hit] = []
    norm_terms = [_normalize(t) for t in terms if t]
    for chunk in corpus:
        score = 0
        content = _normalize(chunk.text or '')
        content_words = [w for w in content.split() if w]

        for term in norm_terms:
            if not term:
                continue
            if term in content:
                score += 2
                continue
            matched = False
            for w in content_words:
                if w.startswith(term) or term.startswith(w):
                    score += 1
                    matched = True
                    break
            if matched:
                continue

        if score > 0:
            hits.append(Hit(chunk, score))

    hits.sort(key=lambda h: h.score, reverse=True)
    return hits[:5]

File: Java/KeywordRetriever.java
package mu.cse.rag.clean;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

public class KeywordRetriever implements Retriever {
    @Override
    public List<Hit> retrieve(List<String> terms, List<Chunk> corpus) {
        List<Hit> hits = new ArrayList<>();
        
        for (Chunk chunk : corpus) {
            int score = 0;
            String content = chunk.text.toLowerCase(Locale.forLanguageTag("tr-TR")); 
            
            for (String term : terms) {
                if (content.contains(term)) {
                    score += 1;
                }
            }
            if (score > 0) {
                hits.add(new Hit(chunk, score));
            }
        }
        hits.sort((h1, h2) -> Integer.compare(h2.score, h1.score)); //Sort by score (descending)
        return hits.stream().limit(5).collect(Collectors.toCollection(ArrayList::new)); //Return top 5 hits
    }
}